### PR TITLE
Be able to decrease background thread's CPU priority when creating database backup

### PR DESCRIPTION
**Pull request by cheng-chang**
When creating a database backup, the background threads will not only consume IO resources by copying files, but also consuming CPU such as by computing checksums. During peak times, the CPU consumption by the background threads might affect online queries.

This PR makes it possible to decrease CPU priority of these threads when creating a new backup.

Test Plan:
make check

### CODE REVIEW

File: include/rocksdb/utilities/backupable_db.h [hunk 1]
```diff
@@ -142,6 +142,14 @@ struct BackupableDBOptions {
   }
 };
 
+struct CreateBackupOptions {
+  // If false, background_thread_cpu_priority is ignored.
+  bool enable_update_background_thread_cpu_priority = false;
+  // Currently only effective on Linux, value ranging from -20 to 19,
+  // 19 is the lowest priority.
+  int background_thread_cpu_priority = 0;
```
Comment (siying) [inline include/rocksdb/utilities/backupable_db.h]: Hmm.. This is too platform dependent and I'm not sure it is a good idea. Can we avoid it for now until we need it? We can always do it for lowest level for now. If one day we need it, we may consider what the option should be.

File: include/rocksdb/utilities/backupable_db.h [hunk 2]
```diff
@@ -262,18 +270,20 @@ class BackupEngine {
   // avoid losing unflushed key/value pairs from the memtable.
   virtual Status CreateNewBackupWithMetadata(
       DB* db, const std::string& app_metadata, bool flush_before_backup = false,
-      std::function<void()> progress_callback = []() {}) = 0;
+      std::function<void()> progress_callback = []() {},
+      const CreateBackupOptions& opt = CreateBackupOptions()) = 0;
```
Comment (siying) [inline include/rocksdb/utilities/backupable_db.h]: My preference is to create a new function
`virtual Status CreateNewBackupWithMetadata(const CreateBackupOptions& opt, DB* db, const std::string& app_metadata)` and put all other arguments into `CreateBackupOptions` and keep the current one by doing a simple translation.

File: utilities/backupable/backupable_db.cc [hunk 3]
```diff
@@ -730,8 +731,22 @@ Status BackupEngineImpl::Initialize() {
...
+    threads_.emplace_back([&mu, &cv, &size, t, this]() {
+      auto id = port::GetCurrentThreadId();
+      {
+        MutexLock l(&mu);
+        thread_ids_[t] = id;
+        size++;
+        if (size == thread_ids_.size()) {
+          cv.SignalAll();
+        }
+      }
```
Comment (siying) [inline utilities/backupable/backupable_db.cc]: This feels complicated. Is it possible to pass `options.enable_update_background_thread_cpu_priority` all the way to here and it just call `port::SetCpuPriority(0, 19);`?

File: port/win/port_win.h [hunk 4]
```diff
@@ -336,6 +336,12 @@ extern int GetMaxOpenFiles();
 std::string utf16_to_utf8(const std::wstring& utf16);
 std::wstring utf8_to_utf16(const std::string& utf8);
 
+using ThreadId = int;
+
+extern void SetCpuPriority(ThreadId id, int priority);
```
Comment (siying) [inline port/win/port_win.h]: `priority` is not platform independent enough, so it is not clear what it means for Windows and other platform. This is probably another reason we should consider to only use priority 19 for now on Linux and avoid this `priority` argument.

File: include/rocksdb/utilities/backupable_db.h (hunk 1 above)
Comment (cheng-chang) [reply include/rocksdb/utilities/backupable_db.h]: According to the person who requests this feature, for each backup, he wants to be able to update both sched type (such as SCHED_IDLE) and priority. I think this requirement is specific for Linux.

File: port/win/port_win.h (hunk 4 above)
Comment (cheng-chang) [reply port/win/port_win.h]: I originally created a function called `SetLowestCpuPriority` which sets the priority to `19`. But afterwards, I find that the user needs to update priority for each backup, so I changed to the current signature. I think the user's feature request is specifically for Linux, I doubt whether we want to make it portable.

File: utilities/backupable/backupable_db.cc (hunk 3 above)
Comment (cheng-chang) [reply utilities/backupable/backupable_db.cc]: these threads are set up during initialization of `BackupEngine`, `options` is passed in for each `CreateNewBackup` after construction of `BackupEngine`,  so I don't think it's possible to pass in the options here.

But once we switch to using [pthread_setschedparam](https://linux.die.net/man/3/pthread_setschedparam) to set both sched type and priority, we don't need the thread ids, and can get rid of this logic here.

File: include/rocksdb/utilities/backupable_db.h (hunk 1 above)
Comment (cheng-chang) [reply include/rocksdb/utilities/backupable_db.h]: created a new enum `CpuPriority` to be platform independent

File: port/win/port_win.h (hunk 4 above)
Comment (cheng-chang) [reply port/win/port_win.h]: created `CpuPriority` enum to be platform neutral

File: include/rocksdb/utilities/backupable_db.h [hunk 5]
```diff
@@ -142,6 +142,20 @@ struct BackupableDBOptions {
...
+  // If write-ahead logs are disabled, set flush_before_backup=true to
+  // avoid losing unflushed key/value pairs from the memtable.
+  bool flush_before_backup = false;
+
+  // Callback for reporting progress.
+  std::function<void()> progress_callback = []() {};
+
+  // If false, background_thread_cpu_priority is ignored.
+  bool enable_update_background_thread_cpu_priority = false;
+  port::CpuPriority background_thread_cpu_priority = port::CpuPriority::kNormal;
```
Comment (siying) [inline include/rocksdb/utilities/backupable_db.h]: CpuPriority  doesn't have to belong to port as it is platform independent. I think we can place it simply in options.h.

Comment (cheng-chang) [reply include/rocksdb/utilities/backupable_db.h]: I thought the `port` module needs to be independent of rocks' public include files. But I saw `port_win.h` includes `options.h`, so I'll `CpuPriority` to `options.h`

File: include/rocksdb/utilities/backupable_db.h [hunk 6]
```diff
@@ -256,24 +275,36 @@ class BackupEngine {
...
+  }
 
   // Captures the state of the database in the latest backup
   // NOT a thread safe call
-  // Flush will always trigger if 2PC is enabled.
-  // If write-ahead logs are disabled, set flush_before_backup=true to
-  // avoid losing unflushed key/value pairs from the memtable.
-  virtual Status CreateNewBackup(DB* db, bool flush_before_backup = false,
+  virtual Status CreateNewBackup(
+      DB* db, const CreateBackupOptions& options = CreateBackupOptions()) {
```
Comment (siying) [inline include/rocksdb/utilities/backupable_db.h]: Can we skip the default and force users to provide an option, like all other options? Also, We tend to put options before other arguments, so it is probably to be better to switch between db and options.

Comment (cheng-chang) [reply include/rocksdb/utilities/backupable_db.h]: I followed https://github.com/facebook/rocksdb/pull/6602/files#diff-d6f7c7da4019d974b2c3c68b4ff2ba84R116 to be consistent

Comment (siying) [reply include/rocksdb/utilities/backupable_db.h]: Can we change...

Comment (cheng-chang) [reply include/rocksdb/utilities/backupable_db.h]: I'll change both

Comment (siying) [reply include/rocksdb/utilities/backupable_db.h]: Please put changing the other function in a separate PR if you want.  I think there is no urgency to do that though.

**Comment by siying** [comment]
"don't remember" => "don't forget"

**Comment by cheng-chang** [comment]
@siying PTAL

**Comment by cheng-chang** [comment]
@siying I'll create another PR for the test.

**Comment by cheng-chang** [comment]
@siying I added some tests to `backupable_db_test.cc`, PTAL.

**Comment by cheng-chang** [comment]
One travis test failed with `IO error: While write random read/write file at offset 251: /tmp/rocksdbtest-1000/rocksdbtest-1000/random_rw_file_rand_9547206656888250462: Disk quota exceeded`, not related to the PR.

File: db/db_impl/db_impl_compaction_flush.cc [hunk 7]
```diff
@@ -2876,11 +2877,11 @@ Status DBImpl::BackgroundCompaction(bool* made_progress,
...
-        if (!io_s.ok()) {
-           error_handler_.SetBGError(io_s, BackgroundErrorReason::kCompaction);
-        } else {
-          error_handler_.SetBGError(status, BackgroundErrorReason::kCompaction);
-        }
+    if (!io_s.ok()) {
+      error_handler_.SetBGError(io_s, BackgroundErrorReason::kCompaction);
+    } else {
+      error_handler_.SetBGError(status, BackgroundErrorReason::kCompaction);
+    }
```
Comment (siying) [inline db/db_impl/db_impl_compaction_flush.cc]: I know those formatting changes probably are generated by "make format". The suggestion is always to separate those changes from this pull request. Including those changes would make this pull request harder to revert (to test or permanently) in the future. Also, it makes it more difficult to blame the code.

Comment (cheng-chang) [reply db/db_impl/db_impl_compaction_flush.cc]: I see, will do it in separate PRs in the future.

**Pull request closed without merging**